process_pr.py

```diff
@@ -1459,7 +1460,50 @@ def process_pr(repo_config, gh, repo, issue, dryRun, cmsbuild_user=None, force=F
             return
 
         print("Processing commits")
+        commit_cache = {k: v for k, v in bot_cache.items() if REGEX_COMMIT_SHA.match(k)}
         if pr.commits < MAX_INITIAL_COMMITS_IN_PR or ok_too_many_commits:
+            seen_commits_cnt = len(commit_cache)
+            if all_commits.totalCount < seen_commits_cnt:
+                print(
+                    "Number of commits in cache ({0}) is greater than the number of commits in PR ({1}) - possible squash detected".format(
+                        seen_commits_cnt, all_commits.totalCount
+                    )
+                )
+
+                last_seen_commit_sha = sorted(
+                    [(k, commit_cache[k]["time"]) for k in commit_cache], key=lambda q: q[1]
+                )[-1][0]
+                new_head_commit_sha = pr.head.sha
+                base_commit_sha = pr.base.sha
+
+                changes_before_squash = repo.compare(base_commit_sha, last_seen_commit_sha)
+                changes_after_squash = repo.compare(base_commit_sha, new_head_commit_sha)
+
+                diff_before_squash = {
+                    (file.filename, file.patch) for file in changes_before_squash.files
+                }
+                diff_after_squash = {
+                    (file.filename, file.patch) for file in changes_after_squash.files
+                }
+
+                if diff_before_squash ^ diff_after_squash:
+                    print("PR diff changed, will not preserve signatures")
+                else:
+                    print("PR diff not changed, preserving signatures")
+                    bot_cache[new_head_commit_sha] = {
+                        "time": bot_cache[last_seen_commit_sha]["time"] + 1,
+                        "files": sorted(
+                            x["filename"]
+                            for x in get_commit(repo.full_name, new_head_commit_sha)["files"]
+                        ),
+                    }
+
+                # Clean removed commits from cache
+                for k in commit_cache:
+                    if not any(k == commit.sha for commit in all_commits):
+                        print("Removing commit {0} from cache".format(k))
+                        del bot_cache[k]
```

Comment:
@iarspider , should we move this block outside the `if all_commits.totalCount < seen_commits_cnt:` condition? e.g. for this PR itself the bot cache has 8 commits while there are only 3 commits ( I think due to force pushes). And I think this block can go in a separate PR as it fixes the issue of invalid commits in bot cache

Reply:
I'm not sure if we will gain anything from moving this block outside of `if` condition - it will just waste CPU cycles for PRs where that condition is false (i.e. any normal PR, without force-pushes). Regarding putting this in a separate PR - well, we can, but that new PR should go in before this one (maybe also use that new PR to move commits under their own key in cache)

Reply:
yes, you are right this block will not do any thing for most PRs but it is useful to clean the commits for those PRs where original commits (before force push) and new commit count is same. e.g. PR with single commit with N force pushes will have N+1 commits in the bot cache.

Ok lets leave the block in this PR